### PR TITLE
Initial implementation of right click drag menu in the shell

### PR DESCRIPTION
**Pull request by yagoulas**
## Purpose

This is a missing feature often used in windows. The purpose is to make it possible to select the desired action when the user drags one or two files with the right button.

Note: this can cause regressions in drag and drop so I prefer to merge it after we branch for the release.

## TODO

- [x] Implement doing a move action if a file is draged in the same driver and a copy action when it is dragged in a different drive
- [x] Show the default action as bold in the popup menu
- [x] Add the new menu resource in all languages
- [ ] Shell extensions support (won't be fixed in this PR)

### CODE REVIEW

File: dll/win32/shell32/droptargets/CFSDropTarget.cpp [hunk 1]
```diff
@@ -180,6 +182,69 @@ BOOL CFSDropTarget::QueryDrop(DWORD dwKeyState, LPDWORD pdwEffect)
     return FALSE;
 }
 
+HRESULT CFSDropTarget::_GetEffectFromMenu(IDataObject *pDataObject, POINTL pt, DWORD *pdwEffect)
+{
+    HMENU hmenu = LoadMenu(shell32_hInstance, MAKEINTRESOURCEW(IDM_DRAGFILE));
+    if (!hmenu)
+        return E_OUTOFMEMORY;
+
+    /* FIXME: We need to support shell extensions here */
```
Comment (gigaherz) [inline dll/win32/shell32/droptargets/CFSDropTarget.cpp]: 👍 I was about to suggest that when I read the comment. :P

Comment (yagoulas) [reply dll/win32/shell32/droptargets/CFSDropTarget.cpp]: Oh I should add it in the todo list

File: dll/win32/shell32/CDefView.cpp [hunk 2]
```diff
@@ -3039,11 +3074,9 @@ HRESULT WINAPI CDefView::Drop(IDataObject* pDataObject, DWORD grfKeyState, POINT
     ImageList_DragLeave(m_hWnd);
     ImageList_EndDrag();
 
-    if ((m_iDragOverItem == -1 || m_pCurDropTarget == NULL) && 
+    if ((IsDropOnSource(NULL) == S_OK) && 
         (*pdwEffect & DROPEFFECT_MOVE) && 
-        /*(GetKeyState(VK_LBUTTON) != 0) &&*/
-        (m_pSourceDataObject.p) && 
-        (SHIsSameObject(pDataObject, m_pSourceDataObject)))
+        (m_grfKeyState & MK_LBUTTON))
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp]: OK so there's no need for SHIsSameObject call anymore?

File: dll/win32/shell32/droptargets/CFSDropTarget.cpp [hunk 3]
```diff
@@ -180,6 +182,69 @@ BOOL CFSDropTarget::QueryDrop(DWORD dwKeyState, LPDWORD pdwEffect)
     return FALSE;
 }
 
+HRESULT CFSDropTarget::_GetEffectFromMenu(IDataObject *pDataObject, POINTL pt, DWORD *pdwEffect)
+{
+    HMENU hmenu = LoadMenu(shell32_hInstance, MAKEINTRESOURCEW(IDM_DRAGFILE));
```
Comment (HBelusca) [inline dll/win32/shell32/droptargets/CFSDropTarget.cpp]: LoadMenuW.

File: dll/win32/shell32/droptargets/CFSDropTarget.cpp [hunk 4]
```diff
@@ -180,6 +182,69 @@ BOOL CFSDropTarget::QueryDrop(DWORD dwKeyState, LPDWORD pdwEffect)
...
+    if (!hmenu)
+        return E_OUTOFMEMORY;
+
+    /* FIXME: We need to support shell extensions here */
+
+    UINT uCommand = TrackPopupMenu(GetSubMenu(hmenu,0),
+                              TPM_LEFTALIGN | TPM_RETURNCMD | TPM_LEFTBUTTON | TPM_RIGHTBUTTON,
```
Comment (HBelusca) [inline dll/win32/shell32/droptargets/CFSDropTarget.cpp]: Indentation looks strange.

File: dll/win32/shell32/droptargets/CFSDropTarget.cpp [hunk 5]
```diff
@@ -180,6 +182,69 @@ BOOL CFSDropTarget::QueryDrop(DWORD dwKeyState, LPDWORD pdwEffect)
...
 
+HRESULT CFSDropTarget::_GetEffectFromMenu(IDataObject *pDataObject, POINTL pt, DWORD *pdwEffect)
+{
+    HMENU hmenu = LoadMenu(shell32_hInstance, MAKEINTRESOURCEW(IDM_DRAGFILE));
+    if (!hmenu)
+        return E_OUTOFMEMORY;
+
+    /* FIXME: We need to support shell extensions here */
+
+    UINT uCommand = TrackPopupMenu(GetSubMenu(hmenu,0),
```
Comment (HBelusca) [inline dll/win32/shell32/droptargets/CFSDropTarget.cpp]: space after "hmenu,"

File: dll/win32/shell32/CDefView.cpp (hunk 2 above)
Comment (yagoulas) [reply dll/win32/shell32/CDefView.cpp]: No, I don't think it was needed in the first place. m_pSourceDataObject is set by the drag source and CDefView::Drop is called at the drop target. This means that in a drag operation there can be two CDefView involved (but it is not nescessary that the source is always a CDefView). The target knows that it is a drag operation in the same control just by seeing that m_pSourceDataObject has a value. This means that a drag operation is active from the target. That's why I think the check for the data object is superfluous.

File: dll/win32/shell32/CDefView.cpp [hunk 6]
```diff
@@ -2423,7 +2424,12 @@ HRESULT STDMETHODCALLTYPE CDefView::GetFocusedItem(int *piItem)
 
 HRESULT STDMETHODCALLTYPE CDefView::GetItemPosition(PCUITEMID_CHILD pidl, POINT *ppt)
 {
-    return E_NOTIMPL;
+    int lvIndex = LV_FindItemByPidl(pidl);
+    if (lvIndex == -1)
+        return E_INVALIDARG;
+
+    m_ListView.GetItemPosition(lvIndex, ppt);
```
Comment (learn-more) [inline dll/win32/shell32/CDefView.cpp]: What happens if this fails (f.e. ppt NULL), should the function still return S_OK?

File: dll/win32/shell32/CDefView.cpp [hunk 7]
```diff
@@ -2717,14 +2737,22 @@ HRESULT STDMETHODCALLTYPE CDefView::GetSelectedObjects(PCUITEMID_CHILD **pidl, U
...
-    FIXME("(%p)->(%p) stub\n", this, drop_target);
-    return E_NOTIMPL;
+    if ((m_iDragOverItem == -1 || m_pCurDropTarget == NULL) && 
+        (m_pSourceDataObject.p))
+    {
+        return S_OK;
+    }
+
+    return S_FALSE;
 }
```
Comment (learn-more) [inline dll/win32/shell32/CDefView.cpp]: We don't need `drop_target` at all?

File: dll/win32/shell32/droptargets/CFSDropTarget.cpp [hunk 8]
```diff
@@ -180,6 +182,69 @@ BOOL CFSDropTarget::QueryDrop(DWORD dwKeyState, LPDWORD pdwEffect)
...
+        return hr;
+
+    PIDLIST_ABSOLUTE pidlFolder;
+    PUITEMID_CHILD *apidl;
+    UINT cidl;
+    hr = SH_GetApidlFromDataObject(pdtobj, &pidlFolder, &apidl, &cidl);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    POINT* apt = static_cast<POINT*>(SHAlloc(cidl * sizeof(POINT)));
```
Comment (learn-more) [inline dll/win32/shell32/droptargets/CFSDropTarget.cpp]: `CComHeapPtr<POINT>`

File: dll/win32/shell32/droptargets/CFSDropTarget.cpp [hunk 9]
```diff
@@ -230,8 +299,28 @@ HRESULT WINAPI CFSDropTarget::Drop(IDataObject *pDataObject,
...
+        HRESULT hr = _GetEffectFromMenu(pDataObject, pt, pdwEffect);
+        if (FAILED_UNEXPECTEDLY(hr) || hr == S_FALSE)
+            return hr;
+    }
+
+    if (*pdwEffect == DROPEFFECT_MOVE && m_site)
+    {
+        CComPtr<IShellFolderView> psfv;
+        HRESULT hr = IUnknown_QueryService(m_site, SID_IFolderView, IID_PPV_ARG(IShellFolderView, &psfv));
+        if (SUCCEEDED(hr) && psfv->IsDropOnSource((IDropTarget*)this) == S_OK)
```
Comment (learn-more) [inline dll/win32/shell32/droptargets/CFSDropTarget.cpp]: Is this explicit cast needed?

File: dll/win32/shell32/lang/en-US.rc [hunk 10]
```diff
@@ -76,6 +76,18 @@ BEGIN
     END
 END
 
+IDM_DRAGFILE MENU
+BEGIN
+    POPUP ""
+    BEGIN
+        MENUITEM "Copy here", IDM_COPYHERE
```
Comment (learn-more) [inline dll/win32/shell32/lang/en-US.rc]: No accelerators?

File: dll/win32/shell32/droptargets/CFSDropTarget.cpp (hunk 1 above)
Comment (learn-more) [reply dll/win32/shell32/droptargets/CFSDropTarget.cpp]: What would this entail?

File: dll/win32/shell32/droptargets/CFSDropTarget.cpp (hunk 9 above)
Comment (yagoulas) [reply dll/win32/shell32/droptargets/CFSDropTarget.cpp]: Not really

File: dll/win32/shell32/CDefView.cpp (hunk 7 above)
Comment (yagoulas) [reply dll/win32/shell32/CDefView.cpp]: To be honest I'm not entirely sure that this what this method is supposed to do in windows.

File: dll/win32/shell32/lang/en-US.rc (hunk 10 above)
Comment (gigaherz) [reply dll/win32/shell32/lang/en-US.rc]: Windows doesn't have them...

explorer_2018-02-17_13-03-04

**Pull request merged**